python/tank/templatekey.py

```diff
@@ -292,31 +304,164 @@ def __init__(self,
+
+        :param value: Value to validate.
 
+        :returns: True is the validation was succesfull, False otherwise.
+        """
         if value is not None:
-            if not (isinstance(value, int) or value.isdigit()):
-                self._last_error = "%s Illegal value %s, expected an Integer" % (self, value)
+            if isinstance(value, basestring):
+                # We have a string, make sure it loosely or strictly matches the format.
+                if self.strict_matching and not self._strictly_matches(value):
+                    return False
+                elif not self.strict_matching and not self._loosely_matches(value):
+                    return False
+            elif not isinstance(value, int):
+                self._last_error = "%s Illegal value '%s', expected an Integer" % (self, value)
                 return False
-            else:
-                return super(IntegerKey, self).validate(value)
+            return super(IntegerKey, self).validate(value)
+        return True
+
+    def _loosely_matches(self, value):
+        """
+        Checks if the value loosely matches. The value loosely matches if it can be turned into an
+        int.
+
+        :param value: String to test
+
+        :returns: True if it loosely matches, False otherwise.
+        """
+        # This is the extent of what was tested before strict matching. We're actually a bit more permissive,
+        # because the user could have specified a format specifier that uses spaces for padding, but isdigit will
+        # fail if spaces are at the beginning of a string, so strip them out.
+        if self._padding_char == ' ':
+            value = value.lstrip()
+        # Is digit is how we tested for a number before strict_matching was introduced, so don't change that behaviour
+        if not value.isdigit():
+            self._last_error = "%s Illegal value '%s', expected an Integer" % (self, value)
+            return False
+        return True
+
+    def _strictly_matches(self, value):
+        """
+        Checks if the value strictly matches the format_spec. A value strictly matches the format
+        spec when it both loosely matches it and is at least as wide as the minimum character
+        length.
+        """
+        error_msg = "%s Illegal value %s, does not match format spec '%s'" % (self, value, self.format_spec)
+        # If there are more characters than the maximum size, we should have a non zero positive number
+        if len(value) > self._minimum_width:
+            if not self._NON_ZERO_POSITIVE_INTEGER_RE.match(value):
+                self._last_error = error_msg
+                return False
+            return True
+
+        if len(value) < self._minimum_width:
+            self._last_error = error_msg
+            return False
+
+        matches = self._strict_validation_re.match(value)
+        if not matches:
+            self._last_error = error_msg
+            return False
```

Comment:
I'll add some comments to better reflect that structure. Maybe it will be less confusing?